suite/tests/client-interface/drbbdup-meta-label-bb-test.dll.c

```diff
@@ -0,0 +1,321 @@
+    CHECK(res == DRBBDUP_SUCCESS, "failed to check whether instr is first non-label");
+    if (is_cur_all_meta)
+        CHECK(!is_first_nonlabel, "no non-label should be available");
+    else
+        CHECK(is_first_nonlabel, "must be first non-label");
+
+    res = drbbdup_is_last_instr(drcontext, instr, &is_last);
+    CHECK(res == DRBBDUP_SUCCESS, "failed to check whether instr is last");
+    CHECK(is_last, "must be last");
+
+    if (is_first && encoding != 0) {
+        instrum_called = true;
+        dr_insert_clean_call(drcontext, bb, where, print_case, false, 1,
+                             OPND_CREATE_INTPTR(encoding));
+    }
+    return DR_EMIT_DEFAULT;
+}
+
+static void
+event_exit(void)
+{
+    drbbdup_status_t res;
+
+    res = drbbdup_exit();
+    CHECK(res == DRBBDUP_SUCCESS, "drbbdup exit failed");
+
+    CHECK(orig_analysis_called, "orig analysis was not done");
+    CHECK(default_analysis_called, "default analysis was not done");
+    CHECK(case1_analysis_called, "case 1 analysis was not done");
+    CHECK(case2_analysis_called, "case 2 analysis was not done");
+
+    CHECK(orig_analysis_destroy_called, "orig analysis was not destroyed");
+    CHECK(case1_analysis_destroy_called, "case 1 analysis was not destroyed");
+    CHECK(case2_analysis_destroy_called, "case 2 analysis was not destroyed");
+
+    CHECK(instrum_called, "instrumentation was not inserted");
+    CHECK(encountered_all_meta, "encountered bb consisting solely of meta instructions");
+
+    bool app2app_res = drmgr_unregister_bb_app2app_event(remove_app_instr);
+    CHECK(app2app_res, "failed to unregister app2app event");
+
+    drmgr_exit();
+}
+
+DR_EXPORT void
+dr_init(client_id_t id)
+{
+    drmgr_init();
+
+    drbbdup_options_t opts = { 0 };
+    opts.struct_size = sizeof(drbbdup_options_t);
+    opts.set_up_bb_dups = set_up_bb_dups;
+    opts.insert_encode = insert_encode;
+    opts.analyze_orig = orig_analyse_bb;
+    opts.destroy_orig_analysis = destroy_orig_analysis;
+    opts.analyze_case_ex = analyse_bb;
+    opts.destroy_case_analysis = destroy_analysis;
+    opts.instrument_instr_ex = instrument_instr;
+    opts.runtime_case_opnd = OPND_CREATE_ABSMEM(&encode_val, OPSZ_PTR);
+    /* Though single-threaded, we sanity-check the atomic load feature. */
+    opts.atomic_load_encoding = true;
+    opts.user_data = NULL;
+    opts.non_default_case_limit = 2;
+    opts.is_stat_enabled = false;
+
+    drmgr_priority_t app2app_priority = { sizeof(app2app_priority),
+                                          TEST_PRIORITY_APP2APP_NAME, NULL, NULL,
+                                          TEST_PRIORITY_APP2APP };
+
+    bool app2app_res =
+        drmgr_register_bb_app2app_event(remove_app_instr, &app2app_priority);
+    CHECK(app2app_res, "app2app failed");
+
+    drbbdup_status_t res = drbbdup_init(&opts);
+    CHECK(res == DRBBDUP_SUCCESS, "drbbdup init failed");
+    dr_register_exit_event(event_exit);
```

Comment:
The whole file is very similar to the other 2 new tests here.  Is there a way to share code?  That would also make it easier to understand these tests, to highlight what matters among the hundreds of identical lines.

Another option would be to merge all 3 tests into one test and replace the first app nop with nothing, the 2nd with a label, and the 3rd with a meta nop, covering all the cases.  Presumably the couple of other changes throughout the file can be updated to accomodate that.  I think that would make it easier to maintain than these 3 separate mostly-duplicated files.   I vote for that.  WDYT?

Reply:
So my rationale was to keep tests as self-contained as possible, but you are right in saying there is a lot of code duplication.

Reply:
I'll see what I can do.

Reply:
I shortened the test-cases so that they are easier to read.